### PR TITLE
IOTA Palantír Messaging (IPM) reference implementation

### PR DESCRIPTION
**Pull request by yillkid**
## Indroduction

This is a swarm-node extension that refer to implement of [IOTA Palantír Messaging (IPM)](https://blog.florence.chat/introducing-iota-palant%C3%ADr-messaging-ipm-c599ed6d2191).

## Installation
* Copy `ipm` module to swarm-node extensions directory
`$ cp -rf ipm <iota-swarm-node>/extensions/`
* Run the swarm-node service
`$ python server.py`

## API Resersnce

* [new_claim](https://hackmd.io/s/SkThdNmUm): Isse a new IPM claim
* [get_all_claims_in_channel](https://hackmd.io/s/ry-8g8nNQ): Get all claims in channel

## How to used
A simple issuer request example as belowm the field `channel` and `next_channel` are all encrypted with the public key of IPM receiver.

```shell
curl http://HOST_URL \
  -X POST \
  -H 'Content-Type: application/json' \
  -d '{"next_channel":"F9JbXSfbvNfyIDLoS8HzbtiPCuIqNlKzBxnIG9wsF5GL8gQWoCIHhnTtuXggL3dhyCky0DBaZ2Rc\nHxL7bO4opAuwZ8UwOV6LeYC8ERRLfRjHWJaZ6rlzV/bYQnPqPmFjgx8Po21N94+IcMm2P6n5lqRg\nMPBs7wVXePjw/cZqJY4=\n","uuid":"V9TCFLAOGGTAQATTJBLABAG9WY","extension":"ipm","sign":"CjkwfY3WzqsBZlSbU5SbJSaum0z24fs+wc7/trzfcoeC0TsfbBaUyANZd8cqRc/LJIKuBjguia6C\n29ykWSLfbnaILXpiTbeku3So3nQqPEA6UMN83dcXMsjN5HIlDkQlf0OueFHmROttCQJQXUKVDSM9\nny3hhvp384cvrGHRLI0=\n","command":"new_claim","msg":"XevYDoGUoRS6G2mVgZvGwllSifBc/gjiSgxAzgdjzA2bC6SwKHXzmAT2f3yYQAFpHKcg3rkMnUwc\n1jTzTxbxA/w2yuDZFj6Lw/tYlrJwy5or48CnDvS5cHyKwFU9h8aZVldaTwGf51CK8Jm2NsfMUKqB\ntDsIYUvAiZjkg2DnNsI=\n","channel":"QYSUOGBPLZTQIS9QUPRCLMPUPW"}'
```

## Demo

A complete workflow demonstration could be found on [TangleID-IPM-sample](https://github.com/yillkid/TangleID-IPM-sample).

### CODE REVIEW

File: config.py [hunk 1]
```diff
@@ -1,2 +1,2 @@
-FULLNODE = "http://node1.puyuma.org:14266"
+FULLNODE = "http://node01.testnet.iotatoken.nl:16265"
```
Comment (jserv) [inline config.py]: Why would you change this?

File: extensions/README.md [hunk 2]
```diff
@@ -1,4 +1,26 @@
 # How to write extensions for IOTA Swarm node
 
-It is straighforard to implement extensions for IOTA Swarm nodes, allowing
-operations being taken regardless of fullnodes.
+This is a guide for how to write a swarm node extension.
```
Comment (jserv) [inline extensions/README.md]: Send another pull request to improve `extensions/README.md` with the aid of Spin.

File: config.py (hunk 1 above)
Comment (yillkid) [reply config.py]: A syncing issue in our testnet node because some neighbors is down, it's a workaround will change back soon.

File: extensions/ipm/main.py [hunk 3]
```diff
@@ -0,0 +1,43 @@
+#import time
+import json
+from swarm_node import get_tips, send_transfer, \
+    find_transactions_by_address
+
+def load(data):
+    result = ""
+
+    print "Hello get data " + str(data)
```
Comment (jserv) [inline extensions/ipm/main.py]: Never place such ugly message dump.

**Comment by jserv** [comment]
@yillkid, You have to summarize what IPM does instead of putting a hyperlink.

**Comment by jserv** [comment]
Since [Chiota](https://github.com/Noc2/Chiota) was meant to adopt IPM, will we consider to take [NTRU Key Exchange for IOTA](https://github.com/Noc2/Chiota/wiki/NTRU-Key-Exchange-for-IOTA)?

File: config.py (hunk 1 above)
Comment (jserv) [reply config.py]: Try to minimize the necessary changes. Explain the crucial parts always.

**Comment by jserv** [comment]
According to the design concepts written by David Hawig, IOTA Palantír Messaging (IPM) consists of the following implementation considerations:
* NTRU Key Exchange
* Only the people participating (with keys) in the chat have access to the next chat address. The incrementing part assures that the chat address is changed even though everyone posts constantly the exact same message.
* File Transfer
* The Long-term Storage

However, this pull request is far away from being IPM-compliant. Don't call it IPM at the moment. It is just hack inspired by IPM.

File: extensions/ipm/main.py [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+    response = send_transfer(tag, json.dumps(data), address, 0, dict_tips, debug=0)
+
+    return str(response)
+
+def get_all_claims_in_channel(data):
+    data = json.loads(data)
+    address = data['channel'].ljust(81,'9')
+    print "Hello address = " + str(address)
+    
+    result = find_transactions_by_address(address)
```
Comment (jserv) [inline extensions/ipm/main.py]: `find_transactions_by_address` does not really make sense for this use case due to its inefficiency.

File: server.py [hunk 5]
```diff
@@ -44,25 +43,17 @@ def do_POST(self):
...
-            if 'debug' not in request_command:
-                debug = 0
-            else:
-                debug = int(request_command['debug'])
-
-            dict_tips = get_tips(int(request_command['tips_type']))
-            result = send_transfer(
-                request_command['tag'], request_command['message'], request_command['address'], int(
-                    request_command['value']), dict_tips, debug)
+        if 'extension' in request_command:
```
Comment (jserv) [inline server.py]: Send another pull request to refactor server code first.

**Comment by yillkid** [comment]
@jserv 

- Change interpreter version in the test suite is necessary because of swarm-node [switch to python3](https://github.com/yillkid/iota-swarm-node/commit/64aca1cda71e7b2f8c54ba2ec172e6eb02a9634b) already.
- Also, increase the service start waiting time from 3 to 10 for some low-level device, but it's a confusing magic number, I will send another pull request to improve that.
- [NTRU Key Exchange for IOTA](https://github.com/Noc2/Chiota/wiki/NTRU-Key-Exchange-for-IOTA) is the other important reason we evaluate to use IPM, but it's not proven to be secure, swarm-node will keep using IOTA one-time signature on value transaction before we complete evaluation of other solutions.

File: server.py [hunk 6]
```diff
@@ -77,4 +68,24 @@ def http_server():
...
+
+    if request_command['command'] == "generate_address":
+        result = generate_address()
+    elif request_command['command'] == "get_tips":
+        result = get_tips(int(request_command['type']))
+    elif request_command['command'] == "send_transfer":
+        if 'debug' not in request_command:
+            debug = 0
+        else:
+            debug = int(request_command['debug'])
```
Comment (ender503) [inline server.py]: Maybe it could be changed to following code.
Give the `debug`  variable default value instead using if-else clause.
```python
            debug = 0
            if 'debug' in request_command:
                debug = int(request_command['debug'])
```

**Comment by yillkid** [comment]
Will close the pull request and divide it into the minimizing necessary changes.
